libcudacxx/include/cuda/std/detail/libcxx/include/cstdlib

```diff
@@ -98,17 +98,22 @@ void *aligned_alloc(size_t alignment, size_t size);                       // C11
 #pragma GCC system_header
 #endif
 
-#ifdef __GNUC__
+// Old nvcc only supports __builtin_unreachable on host
+#if defined(_LIBCUDACXX_COMPILER_NVCC) && _LIBCUDACXX_CUDACC_VER < 1103000
+#ifdef __CUDA_ARCH__
+#  define _LIBCUDACXX_UNREACHABLE() __trap()
```

Comment:
I'm not sure about this choice. Trap is going to generate instructions even in release builds to guarantee certain behavior, whereas the builtin unreachable would lead to UB and used to inform the compiler. Would something like `assume(false)` be a better match? It should not affect codegen in release builds.

Reply:
How about `_LIBCUDACXX_ASSERT(false)` ?

There is another commit dealing with assume

Reply:
Does `_LIBCUDACXX_ASSERT` just expand to normal `assert()`?

The presence of an `assert()` in device code (even unexecuted) can impact performance. It doesn't look like the compiler optimizes away an `assert(false)` either: https://godbolt.org/z/oGGeYYWb8

I think the only thing to do is to leave this definition empty for CUDA < 11.3.

Reply:
That's only because the `assert` is removed all together before the compiler even sees it. 

People have complained in the past about the presence of asserts https://github.com/NVIDIA/libcudacxx/issues/133

Reply:
How about expanding it to something that's obvious UB, like dereferencing an always null pointer?

Reply:
We could define an extern function with a descriptive name and pretend to call that?